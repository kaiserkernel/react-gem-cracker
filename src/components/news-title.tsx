import { gql } from 'apollo-server-express';
import Router from 'next/router';
import * as React from 'react';
import { graphql } from 'react-apollo';
import { parse } from 'url';

import { upvoteNewsItemMutation } from '../data/mutations/upvote-news-item-mutation';

export interface INewsTitleProps extends INewsItemOwnProps {
  upvoteNewsItem: (id: number) => void;
}

export interface INewsItemOwnProps {
  id: number;
  isRankVisible?: boolean;
  isUpvoteVisible?: boolean;
  rank?: number;
  title: string;
  url?: string;
  upvoted?: boolean;
}

export const newsTitleFragment = `
  fragment NewsTitle on NewsItem {
    id
    title
    url
    upvoted
  }
`;

export function NewsTitleView(props: INewsTitleProps): JSX.Element {
  const { id, isRankVisible, isUpvoteVisible, rank, title, upvoted, url } = props;

  return (
    <tr className="athing">
      <td style={{ textAlign: 'right', verticalAlign: 'top' }} className="title">
        <span className="rank">{isRankVisible && `${rank}.`}</span>
      </td>
      <td style={{ verticalAlign: 'top' }} className="votelinks">
        <div style={{ textAlign: 'center' }}>
          {isUpvoteVisible && (
            <a className={upvoted ? 'nosee' : ' '} onClick={() => props.upvoteNewsItem(id)}>
              <div className="votearrow" title="upvote" />
            </a>
          )}
        </div>
      </td>
      <td className="title">
        <a className="storylink" href={url || `item?id=${id}`}>
          {title}
        </a>
        {url && (
          <span className="sitebit comhead">
            {' '}
            (
            <a href={`from?site=${parse(url).hostname}`}>
              <span className="sitestr">{parse(url).hostname}</span>
            </a>
            )
          </span>
        )}
      </td>
    </tr>
  );
}

NewsTitleView.defaultProps = {
  isRankVisible: true,
  isUpvoteVisible: true,
};

export const NewsTitle = graphql<INewsItemOwnProps, {}, {}, INewsTitleProps>(
  gql(upvoteNewsItemMutation),
  {
    props({ ownProps, mutate }) {
      return {
        ...ownProps,
        upvoteNewsItem: (id: number) =>
          mutate!({
            variables: { id },
          })
            // .then(() => Router.push(`/login?id=${id}&password=${password}`))
            .catch(() => Router.push('/login', `/vote?id=${id}&how=up&goto=news`)),
      };
    },
  }
)(NewsTitleView);
