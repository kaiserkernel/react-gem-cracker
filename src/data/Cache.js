// Interface: Commentable (Object can be commented on) comments, commentCount, commenter
// Interface: Voteable (Object can be voted on) upvotes, upvoteCount, hidden, hiddenCount,
// Data Type: Comment, can be on a news item or another comment
// Every time an upvote/downvote/comment is made, update the count

import LRU from 'lru-cache';

import {
  Feed,
} from './models';

// The cache is a singleton

export function warmCache() {
  // Fetch the front pages
  Feed.getForType('HOT', 30, 0);
  Feed.getForType('NEW', 30, 0);
  Feed.getForType('TOP', 30, 0);

  // Run every 15 mins
  setTimeout(warmCache, 1000 * 60 * 15);
}

class Cache {
  isReady = false;

  /*                  BEGIN NEWS ITEMS                      */


  getNewsItem(id) {
    // return this.newsItems.find(newsItem => newsItem.id === id);
    return this.newsItemsCache.get(id);
  }
  setNewsItem(newsItem) {
    return this.newsItemsCache.set(newsItem.id, newsItem);
  }

  /*                  END NEWS ITEMS                      */

  /*                   BEGIN USERS                        */


  getUser(id) {
    return this.userCache.get(id);
  }

  getUsers() {
    return this.userCache.dump();
  }

  createUser(user) {
    this.userCache.set(user.id, user);
    return user;
  }

  /*                    END USERS                         */

  /*                   BEGIN CACHES                         */


  newNewsItemsCache = LRU({
    max: 500,
    maxAge: 1000 * 60 * 60, // 60 Minute cache: ms * s * m
  })
  newsItemsCache = LRU({
    max: 1000,
    maxAge: 1000 * 60 * 60, // 60 Minute cache: ms * s * m
  })
  userCache = LRU({
    max: 500,
    maxAge: 1000 * 60 * 60 * 2, // 2 hour cache: ms * s * m
  })

  /*                   END CACHES                         */
}

export default new Cache();
