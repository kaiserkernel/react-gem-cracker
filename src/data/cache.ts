import { debug } from 'debug';
import LRU from 'lru-cache';

import { FeedSingleton, NewsItem } from './models';
import { FeedType } from './models/feed';
import { User } from './models/user';
import { Comment } from './models/comment';

const logger = debug('app:Cache');
logger.log = console.log.bind(console);

// The cache is a singleton

export function warmCache(): void {
  // Fetch the front pages
  FeedSingleton.getForType(FeedType.TOP, 30, 0);
  FeedSingleton.getForType(FeedType.NEW, 30, 0);

  // Run every 15 mins
  setTimeout(warmCache, 1000 * 60 * 15);
}

class Cache {
  isReady = false;

  /*                  BEGIN NEWS ITEMS                      */

  getNewsItem(id: number): NewsItem | undefined {
    return this.newsItemsCache.get(id.toString());
  }

  setNewsItem(id: number, newsItem: NewsItem): boolean {
    return this.newsItemsCache.set(id.toString(), newsItem);
  }

  /*                  END NEWS ITEMS                      */

  /*                   BEGIN USERS                        */

  getUser(id): User | undefined {
    return this.userCache.get(id);
  }

  getUsers(): Array<LRU.Entry<string, User>> {
    return this.userCache.dump();
  }

  setUser(id: string, user: User): User {
    logger(`Cache set user ${user}`);
    this.userCache.set(id, user);
    return user;
  }

  /*                    END USERS                         */

  /*                   BEGIN COMMENTS                        */

  getComment(id: number): Comment | undefined {
    return this.commentCache.get(id.toString());
  }

  setComment(id: number, comment: Comment): Comment {
    this.commentCache.set(comment.id.toString(), comment);
    logger(`Cache set comment ${comment}`);
    return comment;
  }

  /*                    END COMMENTS                         */

  /*                   BEGIN CACHES                         */

  newNewsItemsCache = new LRU<string, NewsItem>({
    max: 500,
    maxAge: 1000 * 60 * 60, // 60 Minute cache: ms * s * m
  });

  newsItemsCache = new LRU<string, NewsItem>({
    max: 1000,
    maxAge: 1000 * 60 * 60, // 60 Minute cache: ms * s * m
  });

  userCache = new LRU<string, User>({
    max: 500,
    maxAge: 1000 * 60 * 60 * 2, // 2 hour cache: ms * s * m
  });

  commentCache = new LRU<string, Comment>({
    max: 5000,
    maxAge: 1000 * 60 * 60 * 1, // 1 hour cache: ms * s * m
  });

  /*                   END CACHES                         */
}

export const cache = new Cache();
