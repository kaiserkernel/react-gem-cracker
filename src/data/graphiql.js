/**  
   * WARNING: THIS USES THE REAL SCHEMA AND DATABASE!
   *
   *   Take care not to cause data loss but the app should be secure
   */

// TODO: Can make this graphql server part of the main server using feature flag
// Example GraphiQL server https://github.com/graphql/graphiql/tree/master/example

import express from 'express';
import bodyParser from 'body-parser';
import {
  graphqlExpress,
  graphiqlExpress,
} from 'apollo-server-express';

import Schema from './Schema';
import {
  APP_PORT,
  graphQLPath,
  graphiQLPath,
  GRAPHQL_URL,
  GRAPHIQL_URL,
} from '../config';
import {
  Comment,
  Feed,
  NewsItem,
  User,
} from './models';


const logger = console; // global.logger;

const app = express();
// app.use(morgan('dev'));

// Mount GraphQL Server middleware
app.use(graphQLPath, bodyParser.json(), graphqlExpress(request => ({
  schema: Schema,
  rootValue: { request },
  context: {
    Comment,
    Feed,
    NewsItem,
    User,
  },
  debug: true,
})));

app.use(graphiQLPath, graphiqlExpress({
  endpointURL: graphQLPath,
}));

app.listen(APP_PORT, () => {
  logger.info(`==> ðŸ“ˆ GraphQL Server on ${GRAPHQL_URL}`);
  logger.info(`==> ðŸŒŽ  Go to ${GRAPHIQL_URL}`);
});
