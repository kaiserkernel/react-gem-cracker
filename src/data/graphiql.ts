// /**
//  * WARNING: THIS USES THE REAL SCHEMA AND DATABASE!
//  *
//  *   Take care not to cause data loss but the app should be secure
//  */

// // TODO: Can make this graphql server part of the main server using feature flag
// // Example GraphiQL server https://github.com/graphql/graphiql/tree/master/example

// import { graphiqlExpress, graphqlExpress } from 'apollo-server-express';
// import * as bodyParser from 'body-parser';
// import * as express from 'express';
// import { Comment, FeedSingleton, NewsItem, User } from './models';
// import { schema } from './schema';
// import { APP_PORT, GRAPHIQL_URL, graphiQLPath, GRAPHQL_URL, graphQLPath } from '../config';

// const logger = console;

// const app = express();

// // Mount GraphQL Server middleware
// app.use(
//   graphQLPath,
//   bodyParser.json(),
//   graphqlExpress(request => ({
//     schema,
//     rootValue: { request },
//     context: {
//       Comment,
//       Feed: FeedSingleton,
//       NewsItem,
//       User,
//     },
//     debug: true,
//   }))
// );

// app.use(
//   graphiQLPath,
//   graphiqlExpress({
//     endpointURL: graphQLPath,
//   })
// );

// app.listen(APP_PORT, () => {
//   logger.info(`==> ðŸ“ˆ GraphQL Server on ${GRAPHQL_URL}`);
//   logger.info(`==> ðŸŒŽ  Go to ${GRAPHIQL_URL}`);
// });
