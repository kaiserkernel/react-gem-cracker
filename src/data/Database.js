import data from './SampleData';
import cache from './Cache';

/*                  BEGIN NEWS ITEMS                      */


export function getNewsItem(id) {
  return data.newsItems.find(newsItem => newsItem.id === id);
}

export function rankNewsItems() {
  // Would implement this somewhere in the real HN system
  // A scheduled job to recalculate post rank and points
  // .sort((a, b) => (a.rank - b.rank))
}

export function createNewsItem(newsItem) {
  data.newsItems.push(newsItem);
  return newsItem;
}

//                  NEWS ITEM MUTATIONS


export function upvoteNewsItem(id, userId) {
  const newsItemData = data.newsItems.find(newsItem => newsItem.id === id);
  newsItemData.upvotes.add(userId);
  newsItemData.upvoteCount += 1;

  return newsItemData;
}

export function downvoteNewsItem(id, userId) {
  const newsItemData = data.newsItems.find(newsItem => newsItem.id === id);
  newsItemData.downvotes.add(userId);
  newsItemData.downvoteCount += 1;

  return newsItemData;
}

/*                  END NEWS ITEMS                      */

/*                     BEGIN FEED                         */


export function getNewNewsItems(first, skip) {
  return data.new.slice(skip, skip + first)
    .map((postId, index) => ({
      ...getNewsItem(postId),
      rank: skip + index + 1,
    }));
}

export function getTopNewsItems(first, skip) {
  return data.newsItems.slice(skip, skip + first);
}

export function getHotNews() {
  return data.newsItems;
}

export function getNewsItems() {
  return data.newsItems;
}

/*                     END FEED                         */

/*                   BEGIN USERS                        */


export function getUser(id) {
  return data.users.find(user => user.id === id);
}

export function getUsers() {
  return data.users;
}

export function createUser(user) {
  data.users.push(user);
  return user;
}

/*                    END USERS                         */
