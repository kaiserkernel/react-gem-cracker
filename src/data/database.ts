import { debug } from 'debug';

import { cache } from './cache';
import { FeedSingleton, User } from './models';
import { sampleData } from './sample-data';
import { NewsItem } from './models/news-item';

const logger = debug('app:Database');
logger.log = console.log.bind(console);

/*                  BEGIN NEWS ITEMS                      */

export function getNewsItem(id: number): NewsItem | undefined {
  return sampleData.newsItems.find((newsItem) => newsItem.id === id);
}

export function createNewsItem(newsItem: NewsItem): NewsItem {
  sampleData.newsItems.push(newsItem);

  return newsItem;
}

//                  NEWS ITEM MUTATIONS

export function upvoteNewsItem(id: number, userId: string): NewsItem | undefined {
  // Upvote the News Item in the DB
  const newsItem = cache.getNewsItem(id);

  if (newsItem && !newsItem.upvotes.includes(userId)) {
    newsItem.upvotes.push(userId);
    newsItem.upvoteCount += 1;
    cache.setNewsItem(id, newsItem);
  }

  return newsItem;
}

export function unvoteNewsItem(id: number, userId: string): NewsItem | undefined {
  const newsItem = cache.getNewsItem(id);

  if (newsItem && !newsItem.upvotes.includes(userId)) {
    newsItem.upvotes.splice(newsItem.upvotes.indexOf(userId), 1);
    newsItem.upvoteCount -= 1;
    cache.setNewsItem(id, newsItem);
  }

  return newsItem;
}

export function hideNewsItem(id: number, userId: string): NewsItem {
  logger(`Hiding News Item ${id} by ${userId}`);

  const newsItem = cache.getNewsItem(id);
  const user = cache.getUser(userId);

  if (user && !user.hides.includes(id) && newsItem && !newsItem.hides.includes(userId)) {
    user.hides.push(id);
    cache.setUser(userId, user);

    newsItem.hides.push(userId);
    cache.setNewsItem(id, newsItem);

    logger(`Hid News Item ${id} by ${userId}`);
  } else {
    throw new Error(`Data error, user has already hidden ${id} by ${userId}`);
  }

  return newsItem;
}

export function submitNewsItem(id: number, newsItem: NewsItem) {
  // Submit the News Item in the DB
  if (cache.setNewsItem(id, newsItem)) {
    FeedSingleton.new.unshift(id);
    FeedSingleton.new.pop();
    return newsItem;
  }

  throw new Error('Unable to submit News Item.');
}

/*                  END NEWS ITEMS                      */

/*                     BEGIN FEED                         */

export function getNewNewsItems(first: number, skip: number) {
  return sampleData.new.slice(skip, skip + first).map((postId, index) => ({
    ...getNewsItem(postId),
    rank: skip + index + 1,
  }));
}

export function getTopNewsItems(first: number, skip: number) {
  return sampleData.newsItems.slice(skip, skip + first);
}

export function getHotNews(): NewsItem[] {
  return sampleData.newsItems;
}

export function getNewsItems(): NewsItem[] {
  return sampleData.newsItems;
}

/*                     END FEED                         */

/*                   BEGIN USERS                        */

export function getUser(id: string) {
  return sampleData.users.find((user) => user.id === id);
}

export function getUsers() {
  return sampleData.users;
}

export function createUser(user: User) {
  sampleData.users.push(user);

  return user;
}

/*                    END USERS                         */
