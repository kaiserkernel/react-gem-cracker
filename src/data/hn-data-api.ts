import { debug } from 'debug';
import * as Firebase from 'firebase';

import { HN_API_URL, HN_API_VERSION, HN_DB_URI } from '../config';
import { cache } from './cache';
import { Comment, FeedSingleton, NewsItem, User } from './models';

const logger = debug('app:HNDataAPI');
logger.log = console.log.bind(console);

if (!Firebase.apps.length) {
  Firebase.initializeApp({ databaseURL: HN_DB_URI });
}

const api = Firebase.database().ref(HN_API_VERSION);

// https://github.com/HackerNews/API

/* BEGIN NEWS ITEMS */

export function fetchNewsItem(id) {
  logger(`Fetching post ${HN_API_URL}/item/${id}.json`);

  return api
    .child(`item/${id}`)
    .once('value')
    .then(postSnapshot => {
      const post = postSnapshot.val();
      if (post !== null) {
        const newsItem = new NewsItem({
          id: post.id,
          creationTime: post.time * 1000,
          commentCount: post.descendants,
          comments: post.kids,
          submitterId: post.by,
          title: post.title,
          upvoteCount: post.score,
          url: post.url,
        });
        cache.setNewsItem(newsItem.id, newsItem);
        logger(`Created Post: ${post.id}`);
        return newsItem;
      }
      throw post;
    })
    .catch(reason => logger(`Fetching post failed: ${reason}`));
}

export function fetchComment(id) {
  logger(`Fetching comment ${HN_API_URL}/item/${id}.json`);

  return api
    .child(`item/${id}`)
    .once('value')
    .then(itemSnapshot => {
      const item = itemSnapshot.val();
      if (item !== null && !item.deleted && !item.dead) {
        const comment = new Comment({
          comments: item.kids,
          creationTime: item.time * 1000,
          id: item.id,
          parent: item.parent,
          submitterId: item.by,
          text: item.text,
        });

        cache.setComment(comment.id, comment);
        logger(`Created Comment: ${item.id}`);
        return comment;
      }
      throw item;
    })
    .catch(reason => logger(`Fetching comment failed: ${reason}`));
}

export function fetchUser(id) {
  logger(`Fetching user ${HN_API_URL}/user/${id}.json`);

  return api
    .child(`user/${id}`)
    .once('value')
    .then(itemSnapshot => {
      const item = itemSnapshot.val();
      if (item !== null && !item.deleted && !item.dead) {
        const user = new User({
          about: item.about,
          creationTime: item.created * 1000,
          id: item.id,
          karma: item.karma,
          posts: item.submitted,
        });

        cache.setUser(user.id, user);
        logger(`Created User: ${item.id}`, item);
        return user;
      }
      throw item;
    })
    .catch(reason => logger(`Fetching user failed: ${reason}`));
}

export function getFeed(feedType) {
  logger(`Fetching /${feedType}stories.json`);

  return api
    .child(`${feedType}stories`)
    .once('value')
    .then(feedSnapshot => feedSnapshot.val())
    .then(feed => feed.filter(newsItem => newsItem !== undefined && newsItem !== null))
    .catch(reason => logger(`Fetching news feed failed: ${reason}`));
}

const rebuildFeed = feedType => {
  setTimeout(rebuildFeed, 1000 * 60 * 15, feedType);

  getFeed(feedType)
    .then(feed =>
      Promise.all(feed.map(id => fetchNewsItem(id))).then(newsItems => {
        logger(newsItems);

        FeedSingleton[`${feedType}NewsItems`] = newsItems.filter(
          newsItem => newsItem !== undefined && newsItem !== null
        );

        FeedSingleton[feedType] = feed;
        logger(`Updated ${feedType} ids`);
      })
    )
    .catch(reason => logger(`Error building feed: ${reason}`));
};

/* END NEWS ITEMS */

/* BEGIN SEED DATA */

export function seedCache(delay) {
  logger(`Waiting ${delay} ms before seeding the app with data.`);

  // Delay seeding the cache so we don't spam in dev
  setTimeout(() => {
    logger('Seeding cache');
    ['top', 'new', 'best', 'show', 'ask', 'job'].forEach(feedType => {
      rebuildFeed(feedType);
    });
  }, delay);
}

/*  END SEED DATA */
