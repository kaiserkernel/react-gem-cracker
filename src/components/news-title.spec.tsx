import { MockedProvider } from '@apollo/react-testing';
import * as React from 'react';
import MockDate from 'mockdate';
// import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { NewsTitle } from './news-title';
import { sampleData } from '../data/sample-data';

MockDate.set(1506022129802);

describe('NewsTitle component', () => {
  // it('shallow renders', () => {
  //   const wrapper = shallow(
  //     <NewsFeed />,
  //   );
  //   expect(wrapper).toMatchSnapshot();
  // });

  it('renders news item properties passed in as props', () => {
    const wrapper = shallow(
      <MockedProvider>
        <NewsTitle {...sampleData.newsItems[0]} isRankVisible={true} rank={1} />
      </MockedProvider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  // it('renders news items passed in as props', () => {
  //   const wrapper = shallow((
  //     <NewsFeed newsItems={data.newsItems} />
  //   ));
  //   expect(wrapper.contains(<div className="unique" />)).to.equal(true);
  // });
});
