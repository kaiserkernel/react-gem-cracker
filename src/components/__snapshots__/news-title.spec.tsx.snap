// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsTitle component renders news item properties passed in as props 1`] = `
<ApolloProvider
  client={
    ApolloClient {
      "cache": InMemoryCache {
        "addTypename": true,
        "cacheKeyRoot": KeyTrie {
          "weakness": true,
        },
        "config": Object {
          "addTypename": true,
          "dataIdFromObject": [Function],
          "fragmentMatcher": HeuristicFragmentMatcher {},
          "freezeResults": false,
          "resultCaching": true,
        },
        "data": DepTrackingCache {
          "data": Object {},
          "depend": [Function],
        },
        "maybeBroadcastWatch": [Function],
        "optimisticData": DepTrackingCache {
          "data": Object {},
          "depend": [Function],
        },
        "silenceBroadcast": false,
        "storeReader": StoreReader {
          "executeSelectionSet": [Function],
          "executeStoreQuery": [Function],
          "executeSubSelectedArray": [Function],
          "freezeResults": false,
        },
        "storeWriter": StoreWriter {},
        "typenameDocumentCache": Map {},
        "watches": Set {},
      },
      "clearStoreCallbacks": Array [],
      "defaultOptions": Object {},
      "disableNetworkFetches": false,
      "link": MockLink {
        "addTypename": true,
        "mockedResponsesByKey": Object {},
      },
      "localState": LocalState {
        "cache": InMemoryCache {
          "addTypename": true,
          "cacheKeyRoot": KeyTrie {
            "weakness": true,
          },
          "config": Object {
            "addTypename": true,
            "dataIdFromObject": [Function],
            "fragmentMatcher": HeuristicFragmentMatcher {},
            "freezeResults": false,
            "resultCaching": true,
          },
          "data": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "maybeBroadcastWatch": [Function],
          "optimisticData": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "silenceBroadcast": false,
          "storeReader": StoreReader {
            "executeSelectionSet": [Function],
            "executeStoreQuery": [Function],
            "executeSubSelectedArray": [Function],
            "freezeResults": false,
          },
          "storeWriter": StoreWriter {},
          "typenameDocumentCache": Map {},
          "watches": Set {},
        },
        "client": [Circular],
      },
      "mutate": [Function],
      "query": [Function],
      "queryDeduplication": true,
      "queryManager": QueryManager {
        "assumeImmutableResults": false,
        "clientAwareness": Object {
          "name": undefined,
          "version": undefined,
        },
        "dataStore": DataStore {
          "cache": InMemoryCache {
            "addTypename": true,
            "cacheKeyRoot": KeyTrie {
              "weakness": true,
            },
            "config": Object {
              "addTypename": true,
              "dataIdFromObject": [Function],
              "fragmentMatcher": HeuristicFragmentMatcher {},
              "freezeResults": false,
              "resultCaching": true,
            },
            "data": DepTrackingCache {
              "data": Object {},
              "depend": [Function],
            },
            "maybeBroadcastWatch": [Function],
            "optimisticData": DepTrackingCache {
              "data": Object {},
              "depend": [Function],
            },
            "silenceBroadcast": false,
            "storeReader": StoreReader {
              "executeSelectionSet": [Function],
              "executeStoreQuery": [Function],
              "executeSubSelectedArray": [Function],
              "freezeResults": false,
            },
            "storeWriter": StoreWriter {},
            "typenameDocumentCache": Map {},
            "watches": Set {},
          },
        },
        "fetchQueryRejectFns": Map {},
        "idCounter": 1,
        "inFlightLinkObservables": Map {},
        "link": MockLink {
          "addTypename": true,
          "mockedResponsesByKey": Object {},
        },
        "localState": LocalState {
          "cache": InMemoryCache {
            "addTypename": true,
            "cacheKeyRoot": KeyTrie {
              "weakness": true,
            },
            "config": Object {
              "addTypename": true,
              "dataIdFromObject": [Function],
              "fragmentMatcher": HeuristicFragmentMatcher {},
              "freezeResults": false,
              "resultCaching": true,
            },
            "data": DepTrackingCache {
              "data": Object {},
              "depend": [Function],
            },
            "maybeBroadcastWatch": [Function],
            "optimisticData": DepTrackingCache {
              "data": Object {},
              "depend": [Function],
            },
            "silenceBroadcast": false,
            "storeReader": StoreReader {
              "executeSelectionSet": [Function],
              "executeStoreQuery": [Function],
              "executeSubSelectedArray": [Function],
              "freezeResults": false,
            },
            "storeWriter": StoreWriter {},
            "typenameDocumentCache": Map {},
            "watches": Set {},
          },
          "client": [Circular],
        },
        "mutationStore": MutationStore {
          "store": Object {},
        },
        "onBroadcast": [Function],
        "pollingInfoByQueryId": Map {},
        "queries": Map {},
        "queryDeduplication": true,
        "queryStore": QueryStore {
          "store": Object {},
        },
        "ssrMode": false,
        "transformCache": WeakMap {},
      },
      "reFetchObservableQueries": [Function],
      "resetStore": [Function],
      "resetStoreCallbacks": Array [],
      "store": DataStore {
        "cache": InMemoryCache {
          "addTypename": true,
          "cacheKeyRoot": KeyTrie {
            "weakness": true,
          },
          "config": Object {
            "addTypename": true,
            "dataIdFromObject": [Function],
            "fragmentMatcher": HeuristicFragmentMatcher {},
            "freezeResults": false,
            "resultCaching": true,
          },
          "data": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "maybeBroadcastWatch": [Function],
          "optimisticData": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "silenceBroadcast": false,
          "storeReader": StoreReader {
            "executeSelectionSet": [Function],
            "executeStoreQuery": [Function],
            "executeSubSelectedArray": [Function],
            "freezeResults": false,
          },
          "storeWriter": StoreWriter {},
          "typenameDocumentCache": Map {},
          "watches": Set {},
        },
      },
      "typeDefs": undefined,
      "version": "2.6.8",
      "watchQuery": [Function],
    }
  }
>
  <NewsTitle
    commentCount={44}
    comments={Array []}
    creationTime={1504740146516}
    hidden={false}
    hiddenCount={0}
    hides={Array []}
    id={1224}
    isRankVisible={true}
    rank={1}
    submitterId="hvo"
    text={null}
    title="“Learning How to Learn,” the most popular course on Coursera"
    upvoteCount={171}
    upvotes={Array []}
    url="https://www.nytimes.com/2017/08/04/education/edlife/learning-how-to-learn-barbara-oakley.html?hpw&rref=education&action=click&pgtype=Homepage&module=well-region&region=bottom-well&WT.nav=bottom-well"
  />
</ApolloProvider>
`;
